lib/elixir/lib/process.ex

```diff
@@ -420,7 +420,13 @@ defmodule Process do
   """
   @spec register(pid | port, atom) :: true
   def register(pid, name) when not name in [nil, false, true] and is_atom(name) do
-    :erlang.register(name, pid)
+    try do
+      :erlang.register(name, pid)
+    rescue
+      error in ArgumentError ->
+        message = Process.RegisterError.message(pid, name, error.message)
+        reraise %{ error | message: message }, System.stacktrace
```

Comment:
We are building an exception and not really using it. We also don't care about the stacktrace, so I would rewrite this as:

```elixir
  def register(pid, name) when not name in [nil, false, true] and is_atom(name) and is_pid(pid) do
    :erlang.register(pid, name)
  catch
    :error, :badarg ->
      message = "..."
      :erlang.error ArgumentError.exception(message), [pid, name]
  end
```

Where `message = "..."` is the message that you have written. This also keep it semantically closer to what Erlang does, without leaking the :erlang.register call in the stacktrace.